### PR TITLE
Read-in-order over query plan

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `read-in-order` optimization on top of query plan. It is enabled by default. Set `query_plan_read_in_order = 0` to use previous AST-based version.

### CODE REVIEW

File: tests/queries/0_stateless/01952_optimize_distributed_group_by_sharding_key.reference
Comment (KochetovNicolai) [inline tests/queries/0_stateless/01952_optimize_distributed_group_by_sharding_key.reference:21]: Changed after `liftUpUnion`

File: tests/queries/0_stateless/02155_read_in_order_max_rows_to_read.sql
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02155_read_in_order_max_rows_to_read.sql:17]: Now, both of this queries works.

File: tests/queries/0_stateless/02377_optimize_sorting_by_input_stream_properties_explain.reference
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02377_optimize_sorting_by_input_stream_properties_explain.reference:15]: From the comment, there should have been `MergingSortedTransform` initially. `MergeSortingTransform` is a full sort.
It was magically fixed.

File: tests/queries/0_stateless/02377_optimize_sorting_by_input_stream_properties_explain.reference
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02377_optimize_sorting_by_input_stream_properties_explain.reference:95]: This is an incorrect property, but it won't be used cause plan is already built.
We should probably fix it later.

File: tests/queries/0_stateless/02381_join_dup_columns_in_plan.reference
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02381_join_dup_columns_in_plan.reference:11]: Changed after `liftUpUnion`

**Comment by KochetovNicolai** [comment]
Tsan should be fixed in #43009

**Comment by kitaisreal** [comment]
Something is wrong with performance tests, probably it does not work for such queries https://s3.amazonaws.com/clickhouse-test-reports/42829/7ac258c2a77fa813052ac67cc67977b05368fa1d/performance_comparison_[4/4]/report.html.
We need to add a lot of additional performance tests to cover all scenarious.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (CurtizJ) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:568]: Is it possible? Shouldn't we build dag just for input columns even without any functions?

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (CurtizJ) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:369]: Need a comment why do we do this if node is not in matches.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (CurtizJ) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:397]: How is it possible to get `nullptr` in `mapped_children`?

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (CurtizJ) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:275]: Does it mean that function is mapped to itself?

**Review by CurtizJ — COMMENTED** [review]
In general ok, but let's wait for review from someone who is more familiar with code near plan optimizations.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:568]: It is possible. This DAG is from SELECT. For example, `SELECT * FROM tab ORDER BY a, b` will not have any expression steps, and DAG will be empty.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:397]: Yes. Will write a comment.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:275]: Yes, but from different DAGs.
I mean, this is the idea - to find equal calculations if DAGs

**Comment by KochetovNicolai** [comment]
Lol, one query from `order_by_read_in_order` perftest became slower cause read-in-order actually did not work for it
```
EXPLAIN PIPELINE
SELECT *
FROM datasets.hits_100m_obfuscated
WHERE UserID = 1988954671305023629
ORDER BY
    CounterID ASC,
    EventDate ASC
LIMIT 100
SETTINGS query_plan_read_in_order = 1

Query id: 7a3efb48-35f5-428d-a1f6-51668b44a779

┌─explain────────────────────────┐
│ (Expression)                   │
│ ExpressionTransform            │
│   (Limit)                      │
│   Limit                        │
│     (Sorting)                  │
│       (Expression)             │
│       ExpressionTransform      │
│         (ReadFromMergeTree)    │
│         MergeTreeInOrder 0 → 1 │
└────────────────────────────────┘

9 rows in set. Elapsed: 0.006 sec. 

ip-172-31-26-174.ec2.internal :) explain pipeline SELECT * FROM datasets.hits_100m_obfuscated WHERE UserID=1988954671305023629 ORDER BY CounterID, EventDate LIMIT 100 settings query_plan_read_in_order=0

EXPLAIN PIPELINE
SELECT *
FROM datasets.hits_100m_obfuscated
WHERE UserID = 1988954671305023629
ORDER BY
    CounterID ASC,
    EventDate ASC
LIMIT 100
SETTINGS query_plan_read_in_order = 0

Query id: 73f61e3e-89ac-488c-97b7-2522428895b7

┌─explain─────────────────────────────┐
│ (Expression)                        │
│ ExpressionTransform                 │
│   (Limit)                           │
│   Limit                             │
│     (Sorting)                       │
│     MergingSortedTransform 16 → 1   │
│       (Expression)                  │
│       ExpressionTransform × 16      │
│         (ReadFromMergeTree)         │
│         MergeTreeInOrder × 16 0 → 1 │
└─────────────────────────────────────┘
```

**Comment by CurtizJ** [comment]
> Lol, one query from order_by_read_in_order perftest became slower cause read-in-order actually did not work for it

@KochetovNicolai 

But in the second case:
```sql
EXPLAIN PIPELINE
SELECT *
FROM datasets.hits_100m_obfuscated
WHERE UserID = 1988954671305023629
ORDER BY
    CounterID ASC,
    EventDate ASC
LIMIT 100
SETTINGS query_plan_read_in_order = 0

Query id: 73f61e3e-89ac-488c-97b7-2522428895b7

┌─explain─────────────────────────────┐
│ (Expression)                        │
│ ExpressionTransform                 │
│   (Limit)                           │
│   Limit                             │
│     (Sorting)                       │
│     MergingSortedTransform 16 → 1   │
│       (Expression)                  │
│       ExpressionTransform × 16      │
│         (ReadFromMergeTree)         │
│         MergeTreeInOrder × 16 0 → 1 │
└─────────────────────────────────────┘

```
 reading in order also worked, because we have `MergeTreeInOrder` and `MergingSortedTransform` instead of `MergeTreeThread`, `MergeSorting` and `PartialSorting`. The difference is in number of threads.

**Comment by CurtizJ** [comment]
The case with `monotonous_order_by` is reasonable. I think the cause is that sorting with `FinishSorting` without or with high `limit` is never faster than regular sorting, because it has the same algorithmic complexity, but quite slower reading method.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:444]: This requires comment. I see that this is related to projections so I am not sure if I got what is going on here.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:30]: Consider to add big comment here, and list cases in which we apply optimization and maybe what we expect from it. This is important because there is lack of documentation, and after all old code will be removed, it will be hard to understand what cases we support and what cases we do not support.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:76]: Why we do not go thought `JOIN`, I expect that there is read in order for hash join.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:93]: std::vector for stack, should be better.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:128]: Need to add some comment, that we go thought injective functions and it is okay.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:111]: This look looks strange, because after it we do not use column if it is not `is_single`. Consider to just check node.children size.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:149]: Need to add comment what this function intend to perform.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:201]: Consider to name function `collectFixedColumns`.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:87]: Maybe better to rename it into FixedNodes, because as I undestand nodes here are not only column, but can be function.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:272]: For large comments consider to use `/** */`.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:277]: Typo `dirrect` `direct`.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:354]: Do not perform allocations inside loop.

Comment (kitaisreal) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:354]: This stack can be extracted from main iteration.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:274]: There can be edge cases when multiple input nodes with same name in DAG. For example when we add constant input in DAG, we add two nodes with same name. I suspect that we can get name clash because of aliases but I could not find simple example, so maybe it is not a problem.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:412]: We need to move allocation out of loop.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:414]: Consider to extract `frame.mapped_children.size()` from loop predicate.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:306]: This function is generally well written, we need to just add some internal comments, for each step that follows documentation above.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:551]: Consider to rename `next_descr_column` into `next_description_column`.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:557]: Consider to rename `descr` into `sort_column_description`.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:671]: Probably if we decide to split this function call in multiple lines, better to have each function argument on the same line.

Comment (kitaisreal) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:671]: Or on the different lines.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (kitaisreal) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:846]: Need to remove this. It should be supported by default from code above, and even better.

**Review by kitaisreal — APPROVED** [review]
Everything is good. Just added small comments for clarification.

File: src/Interpreters/ExpressionAnalyzer.cpp
Comment (KochetovNicolai) [reply src/Interpreters/ExpressionAnalyzer.cpp:1955]: This was a fix of a bug from https://github.com/ClickHouse/ClickHouse/pull/39157
I was reproduced again with a new implementation, and I had to fix it in a different way.

File: src/Interpreters/InterpreterExplainQuery.cpp
Comment (KochetovNicolai) [reply src/Interpreters/InterpreterExplainQuery.cpp:427]: There was a stupid bug that some settings were not applied for `EXPLAIN`.
`InterpreterSelectQuery` copied context and changed settings only for it, and the following `plan.optimize` did not see this settings change.
I don't know if this is applied to `InterpreterSelectQueryAnalyzer` as well, but I think it's better to copy context anyway.

File: src/Processors/QueryPlan/Optimizations/Optimizations.h
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/Optimizations.h:88]: Don't know why, but will do

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:76]: It was not implemented yet
https://github.com/ClickHouse/ClickHouse/pull/40492
And now I think it would be not so difficult indeed.
But I would prefer to do it later.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:93]: But why? afaik std::stack uses std::vector backend

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:201]: Do not agree here. Cause we only add a few more columns based on existing set.

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:306]: I consider moving it somewhere to common interface, but likely will do it later (after aggregation-in-order)

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:846]: I like to have it till next stable release just in case.

**Pull request merged**
